ocw/lib/openstack.py

```diff
@@ -0,0 +1,108 @@
+                self.log_info("Instance termination {} skipped due to dry run mode", server.name)
+            else:
+                self.log_info("Deleting instance {}", server.name)
+                try:
+                    if not self.client().delete_server(
+                            server.name,
+                            wait=False,
+                            timeout=180,
+                            delete_ips=True,  # Delete floating IP address
+                            delete_ip_retry=1):
+                        self.log_err("Failed to delete instance {}", server.name)
+                except OpenStackCloudException as exc:
+                    self.log_warn("Got exception while deleting instance {}: {}", server.name, exc)
+
+    def _cleanup_images(self) -> None:
+        max_days = PCWConfig.get_feature_property('cleanup', 'openstack-max-age-days', self._namespace)
+        now = datetime.now(timezone.utc)
+        try:
+            images = [image for image in self.client().image.images() if "openqa" in image.tags]
+        except OpenStackCloudException as exc:
+            self.log_warn("Got exception while listing images: {}", exc)
+            return
+        self.log_dbg("Found {} images", len(images))
+        for image in images:
+            if (now - parse(image.created_at)).days > max_days:
+                if self.dry_run:
+                    self.log_info("Image deletion {} skipped due to dry run mode", image.name)
+                else:
+                    self.log_info("Deleting image {}", image.name)
+                    try:
+                        if not self.client().delete_image(
+                                image.name,
+                                wait=False,
+                                timeout=3600):
+                            self.log_err("Failed to delete image {}", image.name)
+                    except OpenStackCloudException as exc:
+                        self.log_warn("Got exception while deleting image {}: {}", image.name, exc)
+
+    def _cleanup_keypairs(self) -> None:
+        try:
+            keypairs = [keypair for keypair in self.client().list_keypairs() if keypair.name.startswith("openqa")]
+        except OpenStackCloudException as exc:
+            self.log_warn("Got exception while listing keypairs: {}", exc)
+            return
+        self.log_dbg("Found {} keypairs", len(keypairs))
+        for keypair in keypairs:
+            if self.dry_run:
+                self.log_info("Keypair deletion {} skipped due to dry run mode", keypair.name)
+            else:
+                self.log_info("Deleting keypair {}", keypair.name)
+                try:
+                    if not self.client().delete_keypair(keypair.name):
+                        self.log_err("Failed to delete keypair {}", keypair.name)
+                except OpenStackCloudException as exc:
+                    self.log_warn("Got exception while deleting keypair {}: {}", keypair.name, exc)
```

Comment:
> Why don't we check the key age over here?

The `openstack.compute.v2.keypair.Keypair` does have a `created_at` attribute but our server is not populating it and the Terraform provider doesn't have the capability to add tags to keypairs.

So I'm inclined to remove the whole `_cleanup_keypairs` method and instead of creating a new keypair for each test as done [here](https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/data/publiccloud/terraform/openstack.tf#L26-L30) use a manually created keypair as documented in:
https://registry.terraform.io/providers/terraform-provider-openstack/openstack/latest/docs/resources/compute_keypair_v2

@asmorodskyi @jlausuch WDYT?

Reply:
I see to separate discussions which I don't want to mix into one . Only important point from PCW POV is that we don't want to clean up key-pairs because there is no way to easily understand age of certain key . Did I get this part right ? 

The part about how to handle key creation within openQA tests let's do in corresponding PR to openQA repo

Reply:
> I see to separate discussions which I don't want to mix into one . Only important point from PCW POV is that we don't want to clean up key-pairs because there is no way to easily understand age of certain key . Did I get this part right ?
> 

Yes.

> The part about how to handle key creation within openQA tests let's do in corresponding PR to openQA repo

Agree.  I will remove the whole `cleanup_keypairs()` stuff.

Reply:
I have to call `client.compute.get_keypair(keypair.name)` for each keypair to get the additional info